Code/Framework/AzCore/AzCore/Name/Name.h

```diff
@@ -82,62 +114,101 @@ namespace AZ
-        
-        void SetEmptyString();
-        
+        void SetName(AZStd::string_view name, bool isLiteral = false);
+
+        void SetNameLiteral(AZStd::string_view name);
+
         // This constructor is used by NameDictionary to construct from a dictionary-held NameData instance.
         Name(Internal::NameData* nameData);
 
         static void ScriptConstructor(Name* thisPtr, ScriptDataContext& dc);
 
+        void LinkStaticName(Name** name);
+        void UnlinkStaticName();
+
+        //! If set, this name has been linked to the global name dictionary's static list
+        bool m_linkedToDictionary = false;
+        //! If set, this name can be used for deferred loading
+        bool m_supportsDeferredLoad = false;
+
+        //! The internal hash used by this name.
+        Hash m_hash = 0;
+
         // Points to the string that represents the value of this name.
         // Most of the time this same information is available in m_data, but keeping it here too...
         // - Removes an indirection when accessing the name value.
         // - Allows functions like data() to return an empty string instead of null for empty Name objects.
         AZStd::string_view m_view;
 
-        Hash m_hash;
-
         //! Pointer to NameData in the NameDictionary. This holds both the hash and string pair.
-        AZStd::intrusive_ptr<Internal::NameData> m_data;
+        NameRef m_data;
+
+        //! Describes the begin of the static list of Names that were initialized before the NameDictionary was available.
+        //! On module initialization, these names are linked into the NameDictionary's static pool and created.
+        static Name* s_staticNameBegin;
+        //! Describes the next name entry in the static list of Names, if needed.
+        Name* m_nextName = nullptr;
+        //! Describes the previous name entry in the static list of Names, if needed.
+        Name* m_previousName = nullptr;
     };
 
 } // namespace AZ
 
+//! Defines a cached name literal that describes an AZ::Name. Subsequent calls to this macro will retrieve the cached name from the
+//! dictionary.
+#define AZ_NAME_LITERAL(str)                                                                                                               \
+    (                                                                                                                                      \
+        []() -> AZ::Name                                                                                                                   \
+        {                                                                                                                                  \
+            static AZ::Name nameLiteral(AZ::Name::FromStringLiteral(str));                                                                 \
+            return nameLiteral;                                                                                                            \
```

Comment:
I believe @NicholasVanSickle to support assignment from an AZ_NAME_LITERAL.
if it was just 

```
#define AZ_NAME_LITERAL(str) \
    static AZ::Name nameLiteral(AZ::Name::FromStringLiteral(str));
```
Then assignment would not work.

Furthermore the atomic ref-count operation wouldn't occur in normal circumstances when AZ_NAME_LITERAL would be used, because it would be before the NameDictionary exist and there is no NameData associated with the Name at that point.

Reply:
Yep, the idea is to be able to assign this (and thus treat it like an rvalue). The static holder always needs to maintain one reference to keep the data alive, so you're essentially just paying for a copy of the name - the intention is to replace code like this, that always creates a name:
```
Foo CreateBar()
{
    return Foo(AZ::Name("bar_name"));
}
```
with
```
Foo CreateBar()
{
    return Foo(AZ_NAME_LITERAL("bar_name"));
}
```
which will only do one name dictionary lookup for `bar_name`, then just to a ref count increment going forward.